layer/logged_data/queuing_network_log_destination.py

```diff
@@ -0,0 +1,78 @@
+import queue
+import threading
+from queue import Queue
+from typing import Any, Callable, List, Optional
+
+from layer.clients.logged_data_service import LoggedDataClient
+from layer.logged_data.data_logging_request import DataLoggingRequest
+from layer.logged_data.log_data_destination import LogDataDestination
+from layer.logged_data.retrying_files_sender import RetryingFilesSender
+
+
+WAIT_INTERVAL_SECONDS = 1
+
+
+class QueueingNetworkLogDestination(LogDataDestination):
+    def __init__(self, client: LoggedDataClient) -> None:
+        self._log_data_client = client
+        self._files_storage = RetryingFilesSender()
+
+        self._sending_errors: List[Any] = []
+        self._stop_reading = False
+        self._local_queue: Queue[DataLoggingRequest] = Queue()
+        self._reading_thread = threading.Thread(target=self._send_elem_from_queue)
+        self._reading_thread.start()
+
+    def enqueue(
+        self,
+        func: Callable[[LoggedDataClient], Optional[Any]],
+        data: Optional[Any] = None,
+    ) -> None:
+        self._local_queue.put_nowait(
+            DataLoggingRequest(
+                self._files_storage, lambda: func(self._log_data_client), data
+            )
+        )
+
+    def close(self) -> None:
+        if not self._stop_reading:
+            self._stop_reading = True
+            self._reading_thread.join(timeout=3 * WAIT_INTERVAL_SECONDS)
+            # todo remove this debug line after manual testing
+            if self._reading_thread.is_alive():
+                print("Reading thread still alive, this should not happen.")
+            self._flush()
+            self._files_storage.close()
+            self._print_warning_message_if_errors()
+
+    def _send_elem_from_queue(self) -> None:
+        while not self._stop_reading:
+            try:
+                execution_item = self._local_queue.get(
+                    block=True, timeout=WAIT_INTERVAL_SECONDS
+                )
+                execution_item.execute()
+            except queue.Empty:
+                # no item in the queue, wait for another
+                pass
+            except Exception as ex:
+                self._sending_errors.append(ex)
+
+    def _flush(self) -> None:
+        while not self._local_queue.empty():
+            execution_item = self._local_queue.get(block=False)
+            try:
+                execution_item.execute()
+            except Exception as ex:
+                self._sending_errors.append(ex)
+
+    # todo: naive implementation for now, we should move this to tracker?
+    def _print_warning_message_if_errors(self) -> None:
+        if len(self._sending_errors) == 0:
+            return
+
+        message = "\nWARNING: Layer was unable to log all data because of the following errors:\n"
+        for ex in self._sending_errors:
+            message = message + "Exception: " + str(ex) + "\n"
```

Comment:
nit: 
```suggestion
            message += f"Exception: {ex}\n"
```
Probably works too (haven't tried) and seems neater?

Reply:
thanks, used that syntax